### PR TITLE
Arm64Emitter: Optimize constants with ADRP and ADR

### PR DESCRIPTION
**Pull request by Sonicadvance1**
In a large number of cases we are moving pointers within a 4GB region
and some marginal pointers that are within 1MB.

This is only used in the case that MOVZ can't be used.

NOP padding still occurs after these instructions to ensure that if they
are being used with relocations it will still get padded to a full 4
instruction length.

Not all hardware fuses these and LLVM claims that Cortex beyond A72 even
doesn't, but it'll still be faster.

### CODE REVIEW

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 1]
```diff
@@ -42,12 +42,58 @@ void Arm64Emitter::LoadConstant(vixl::aarch64::Register Reg, uint64_t Constant,
...
+
+  // PC aligned to page
+  uint64_t AlignedPC = PC & ~0xFFFULL;
+
+  // Offset from aligned PC
+  int64_t AlignedOffset = static_cast<int64_t>(Constant) - static_cast<int64_t>(AlignedPC);
+
+  // If the aligned offset is within the 4GB window then we can use ADRP+ADD
+  // and the number of move segments more than 1
+  if (RequiredMoveSegments > 1 && std::abs(AlignedOffset) < 0x7FFF'FFFFLL) {
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:68]: Can we use vixl's `IsInt32` here?

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 2]
```diff
@@ -42,12 +42,58 @@ void Arm64Emitter::LoadConstant(vixl::aarch64::Register Reg, uint64_t Constant,
...
+    // If this is 4k page aligned then we only need ADRP
+    if ((std::abs(AlignedOffset) & 0xFFF) == 0) {
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:70]: abs isn't needed

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 3]
```diff
@@ -42,12 +42,58 @@ void Arm64Emitter::LoadConstant(vixl::aarch64::Register Reg, uint64_t Constant,
...
+      adrp(Reg, AlignedOffset >> 12);
+    }
+    else {
+      // If the constant is within 1MB of PC then we can still use ADR to load in a single instruction
+      // 21-bit signed integer here
+      constexpr int64_t MAX_ADR_OFFSET = 0xF'FFFF;
+      int64_t SmallOffset = static_cast<int64_t>(Constant) - static_cast<int64_t>(PC);
+      if (std::abs(SmallOffset) <= MAX_ADR_OFFSET) {
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:78]: Can we use vixl's `IsInt21` here?

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 4]
```diff
@@ -42,12 +42,58 @@ void Arm64Emitter::LoadConstant(vixl::aarch64::Register Reg, uint64_t Constant,
...
+        adr(Reg, SmallOffset);
+      }
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:80]:
```suggestion
        NumMoves = 1;
      }
```

**Review by skmp — APPROVED** [review]

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp]: Resolved.

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp]: Resolved.

File: External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 1 above)
Comment (Sonicadvance1) [reply External/FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp]: Resolved.

**Pull request merged**